Tasks/Gradle/CodeAnalysis/SonarQube/metrics.ts

```diff
@@ -64,6 +74,65 @@ export class SonarQubeMetrics {
     }
 
     /**
+     * Returns true if the quality gate has warned.
+     * @returns {Promise<boolean>} True if the quality gate has warned, false if the quality gate passed or failed.
+     */
+    public static hasQualityGateWarned(qualityGateStatus:string): boolean {
+        return qualityGateStatus.toUpperCase() == 'WARN';
+    }
+
+    /**
+     * Returns a list of reasons why the quality gate failed. Does not return unviolated failure conditions.
+     * @param analysisDetails JSON-decoded object representing the analysis details
+     * @returns {SonarQubeFailureCondition[]} Array of violated failure conditions
+     */
+    public static getFailedConditions(analysisDetails:any):SonarQubeFailureCondition[] {
+        var result:SonarQubeFailureCondition[] = [];
+        analysisDetails.projectStatus.conditions.forEach((condition:any) => {
+            var sqCondition = condition as SonarQubeFailureCondition;
+
+            // Only add the condition if they caused the quality gate to warn or fail, otherwise skip it
+            if (sqCondition.status.toUpperCase() == 'WARN' || sqCondition.status.toUpperCase() == 'ERROR') {
+                result.push(sqCondition);
+            }
+        });
+        return result;
+    }
+
+    /**
+     * Makes a RESTful API call to get analysis details (e.g. quality gate status)
+     * Returns from the cache if analysis has completed.
+     * @param analysisId (optional) String representing the ID of the analysis to fetch details for.
+     *     If not specified, cached ID will be used or analysisId will be fetched
+     * @returns JSON object representation of the analysis details
+     */
+    public getAnalysisDetails(analysisId?:string):Q.Promise<Object> {
```

Comment:
A small naming issue - methods such as getX suggest that the operation is a fast completing one. However, getAnalysisDetails involves making rest api calls etc. So perhaps fetchAnalysisDetails or queryFor is better?